File: daily-exercise/high_frequency/3sum.go
// 快速排序
func quickSort(left, right int, nums []int) {
    if left >= right {
        return
    }
    low, high := left, right
    pivot := nums[low]
    for left < right {
        for left < right && nums[right] > pivot {
            right --
        }
        nums[left] = nums[right]
        for left < right && nums[left] <= pivot {
            left ++
        }
        nums[right] = nums[left]
    }
    nums[left] = pivot
    quickSort(left+1,high,nums)
    quickSort(low,right-1,nums)
}

func threeSum(nums []int) [][]int {
    n := len(nums)
	if nums == nil || n < 3 {
		return nil
	}
	// 排序
	//sort.Ints(nums)
	left,right := 0,n-1
    quickSort(left,right,nums)

    //三指针
	targetArray := make([][]int,0)
    for point := 0;point < (len(nums)-2);point++ {
        if nums[point] > 0 {
            break
        }
        if point > 0  && nums[point] == nums[point-1] {
            continue
        }
        //双指针
        head,tail := point+1 ,n-1
        for head < tail {
            numSum := nums[point]+nums[head]+nums[tail]
            if numSum < 0 {
                head ++
                for head < tail && nums[head] == nums[head-1] {
                    head ++
                }
            } else if numSum > 0{
                tail --
                for head < tail && nums[tail] == nums[tail+1] {
                    tail --
                }
            } else {
                tmpArray := []int{nums[point], nums[head], nums[tail]}
                targetArray = append(targetArray, tmpArray)
                head ++
                tail --
                for head < tail && nums[head] == nums[head-1] {
                    head ++
                }
                for head < tail && nums[tail] == nums[tail+1] {
                    tail --
                }
            }
        }
    }
    return targetArray
}

File: daily-exercise/high_frequency/3sum.py
# -*- coding: utf-8 -*-
# @Time    : 2022/7/19 9:33
# @Author  : kevin
# @Version : python 3.7
# @Desc    : 三数之和(先排序再双指针)

class Solution:
    def threeSum(self, nums):
        if (not nums or len(nums) < 3):
            return []
        ## 快排
        # def quick_sort(nums, left, right):
        #     if left >= right:
        #         return
        #     可加入随机流程避免最坏情况（完全顺序或逆序）
        #     low, high = left, right
        #     pivot = nums[low]
        #     while left < right:
        #         while left < right and nums[right] > pivot:
        #             right -= 1
        #         nums[left] = nums[right]
        #         while left < right and nums[left] <= pivot:
        #             left += 1
        #         nums[right] = nums[left]
        #     nums[left] = pivot
        #     quick_sort(nums, low, left - 1)
        #     quick_sort(nums, left + 1, high)

        # left, right = 0, len(nums) - 1
        # quick_sort(nums, left, right)
        nums.sort()

        ## 三指针
        target_array = []
        for point in range(0, len(nums) - 2):  # 外层point移动
            if nums[point] > 0:  # point小于head和tail，>0表示没有符合条件的三元组
                break
            if point > 0 and nums[point] == nums[point - 1]:  # 去除point重复情况
                continue
            head, tail = point + 1, len(nums) - 1  # 确定point后，内层定义头尾双指针
            while head < tail:  # 内层双指针循环
                num_sum = nums[point] + nums[head] + nums[tail]  # sum
                if num_sum < 0:  # 移动head
                    head += 1
                    while head < tail and nums[head] == nums[head - 1]:  # 去除head重复情况
                        head += 1
                elif num_sum > 0:  # 移动tail
                    tail -= 1
                    while head < tail and nums[tail] == nums[tail + 1]:  # 去除tail重复情况
                        tail -= 1
                else:  # 符合条件，双向内缩移动
                    target_array.append([nums[point], nums[head], nums[tail]])
                    head += 1
                    tail -= 1
                    # 去除head和tail重复情况
                    while head < tail and nums[head] == nums[head - 1]:
                        head += 1
                    while head < tail and nums[tail] == nums[tail + 1]:
                        tail -= 1

        return target_array

# [-1,0,1,2,-1,-4]
